### PR TITLE
JVM_IR: resolve inline fake overrides before codegen

### PR DESCRIPTION
**Pull request by pyos**
See KT-33054.

### CODE REVIEW

File: compiler/testData/codegen/box/jvm8/kt33054.kt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// TARGET_BACKEND: JVM
+// IGNORE_BACKEND: JVM
+// JVM_TARGET: 1.8
+
+open class A(val x: String) {
+    inline fun f() = if (this is C) this else A(x)
```
Comment (max-kammerer) [inline compiler/testData/codegen/box/jvm8/kt33054.kt]: Could you also add test with inline getter?

Comment (pyos) [reply compiler/testData/codegen/box/jvm8/kt33054.kt]: Something like this? https://github.com/JetBrains/kotlin/compare/08afa065ed01c48604f3620e3a2fa76437aab4fe..62fa1f00f6ecabd181b1f78ff1fdcf18fa005212

Comment (max-kammerer) [reply compiler/testData/codegen/box/jvm8/kt33054.kt]: Yes!

**Comment by max-kammerer** [comment]
Looks like this commit also affects https://youtrack.jetbrains.com/issue/KT-29242, could you also add test from this issue?

**Comment by pyos** [comment]
Isn't that exactly the same code, i.e. `this` in `flatMap` has type `Success` after inlining while the other branch is `Sample`, so from asm's viewpoint the join of those is `Object`?

**Comment by max-kammerer** [comment]
Yes, looks like it's one problem. Actually I want to have test similar to original one reported in issue KT-29242 or KT-33054

**Comment by max-kammerer** [comment]
..in additional to existing one

**Comment by pyos** [comment]
Added copy of the code from KT-29242 as a test.

**Comment by max-kammerer** [comment]
Thank you!

File: compiler/testData/codegen/box/jvm8/kt29242.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
+// TARGET_BACKEND: JVM
+// IGNORE_BACKEND: JVM
+// JVM_TARGET: 1.8
```
Comment (max-kammerer) [inline compiler/testData/codegen/box/jvm8/kt29242.kt]: `JVM_TARGET ` is redundant. We have special test suite that set target to 1.8 in codegenTarget8Jvm8Test. And run configuration for it in https://teamcity.jetbrains.com/buildConfiguration/Kotlin_KotlinPublic_CodegenTestsOnDifferentJDKs?branch=

`TARGET_BACKEND` could be also removed

Comment (max-kammerer) [reply compiler/testData/codegen/box/jvm8/kt29242.kt]: So it's best to convert these tests to boxInline

Comment (pyos) [reply compiler/testData/codegen/box/jvm8/kt29242.kt]: Are you sure this is true for local runs? Without JVM_TARGET, these tests pass on the non-IR backend. (I'm running them as `./gradlew cleanTest :compiler:test --parallel`). The rest of the files in this directory seem to have both JVM_TARGET and TARGET_BACKEND as well.

Comment (pyos) [reply compiler/testData/codegen/box/jvm8/kt29242.kt]: Moved to boxInline,

Comment (max-kammerer) [reply compiler/testData/codegen/box/jvm8/kt29242.kt]: All of these tests are require //JVM_TARGET: 1.8 cause they wouldn't work on default target 1.6. And as result it's implies JVM target restriction.

And yes, you are right. This test would be passed on default target and failed on custom one. 
OK, let's keep tests as is. And after fixing problem in current backend I will convert this tests into boxInline.

Comment (pyos) [reply compiler/testData/codegen/box/jvm8/kt29242.kt]: Moved back to box/jvm8.

**Comment by max-kammerer** [comment]
Thank you!

**Pull request merged**